server/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java

```diff
@@ -73,15 +73,28 @@ public RestStatus status() {
             // on coordinator node. so get the status from cause instead of returning SERVICE_UNAVAILABLE blindly
             return getCause() == null ? RestStatus.SERVICE_UNAVAILABLE : ExceptionsHelper.status(getCause());
         }
-        RestStatus status = shardFailures[0].status();
-        if (shardFailures.length > 1) {
-            for (int i = 1; i < shardFailures.length; i++) {
-                if (shardFailures[i].status().getStatus() >= RestStatus.INTERNAL_SERVER_ERROR.getStatus()) {
-                    status = shardFailures[i].status();
-                }
+        RestStatus status = null;
+        for (ShardSearchFailure shardFailure : shardFailures) {
+            RestStatus shardStatus = shardFailure.status();
+            switch (shardStatus) {
+                // Return if it's an error that can be retried.
+                // These currently take precedence over other status code(s).
+                case RestStatus.BAD_GATEWAY:
+                case RestStatus.SERVICE_UNAVAILABLE:
+                case RestStatus.GATEWAY_TIMEOUT:
+                    return shardStatus;
+
+                // Although these status codes cannot be retried, they should still be tracked because
+                // they're still of interest to us.
+                case RestStatus.INTERNAL_SERVER_ERROR:
+                case RestStatus.NOT_IMPLEMENTED:
+                case RestStatus.HTTP_VERSION_NOT_SUPPORTED:
+                case RestStatus.INSUFFICIENT_STORAGE:
```

Comment:
can you expand on these special cases? The last two are never returned from ES, and the second one only from some very specific places, I wonder why we need to treat it differently. Could we otherwise have some kind of logic like if status == 500 or status > 504 etc. ? Is that the intention?

Reply:
The code that's being replaced tracks for any status code as long as it's >= 500. So I retained them here in the `switch` as well. Since you say that we don't return the last 2, should I get rid of them then?

Reply:
I see, so you have a preference for mentioning each specific status code? I think it's a matter of preference, I have a hard time connecting those descriptions with the actual status code number, hence I have a slight preference for reasoning about error ranges perhaps. It may be me.

Reply:
Since you mentioned that the last 2 are never really returned and the 2nd one only from specific places, I could remove them. I'm now leaning towards this approach. Alternately, I could retain them and simply add a code comment.

Reply:
I can change the code to use actual status codes rather than the `enum`-s. In fact, I'm alright with that. And regarding the status codes that aren't used/returned (the ones you've talked in the first comment), should I get rid of them?